pdns/pdns_recursor.cc

```diff
@@ -4054,13 +4150,44 @@ static void handleTCPClientWritable(int fd, FDMultiplexer::funcparam_t& var)
   else {  // error or EOF
     PacketID tmp(*pid);
     t_fdm->removeWriteFD(fd);
-    string sent;
+    PacketBuffer sent;
+    MT->sendEvent(tmp, &sent);         // we convey error status by sending empty string
+  }
+}
+
+static void TCPIOHandlerWritable(int fd, FDMultiplexer::funcparam_t& var)
+{
+  PacketID* pid = boost::any_cast<PacketID>(&var);
+  assert(pid->tcphandler != nullptr);
+  assert(fd == pid->tcphandler->getDescriptor());
+
+  try {
+    IOState state = pid->tcphandler->tryWrite(pid->outMSG, pid->outPos, pid->outMSG.size());
+    switch (state) {
+    case IOState::Done: {
+      PacketID tmp = *pid;
```

Comment:
I wonder if we could move `*pid` here and in the exception handling below, to prevent a full alloc+copy?